src/toml.ts

```diff
@@ -1,65 +1,175 @@
 import { Section, stringify, parse } from '@ltd/j-toml'
 import { isObject } from './utils/isObject'
-import { IWorkflowNode, IParameters, IWorkflow } from './types'
+import { IWorkflowNode, IParameters, IWorkflow, TomlObjectSchema } from './types'
 
-function nodes2tomltable (nodes: IWorkflowNode[]): Record<string, unknown> {
+function nodes2tomltable (nodes: IWorkflowNode[], tomlSchema4nodes: Record<string, TomlObjectSchema> = {}): Record<string, unknown> {
   const table: Record<string, unknown> = {}
   const track: Record<string, number> = {}
   for (const node of nodes) {
     if (!(node.id in track)) {
       track[node.id] = 0
     }
     track[node.id]++
+    const tomlSchemaOfNode = node.id in tomlSchema4nodes ? tomlSchema4nodes[node.id] : {}
     const section =
       track[node.id] > 1 ? `${node.id}.${track[node.id]}` : node.id
-    const nodeParameters: Record<string, unknown> = {}
-    // TODO make recursive so `items.input.items.hisd: nesting` is also applied
-    Object.entries(node.parameters).forEach(([k, v]) => {
-      if (Array.isArray(v) && v.length > 0 && isObject(v[0])) {
-        // A value that is an array of objects will have each of its objects as a section
-        nodeParameters[k] = v.map(d => Section(d))
-      } else if (isObject(v)) {
-        nodeParameters[k] = Section(v as any)
-      } else {
-        nodeParameters[k] = v
-      }
-    })
+    const nodeParameters: Record<string, unknown> = parameters2toml(node.parameters, tomlSchemaOfNode)
     table[section] = Section(nodeParameters as any)
   }
   return table
 }
 
+function parameters2toml (parameters: IParameters, tomlSchema: TomlObjectSchema): any {
+  const tomledParameters: Record<string, unknown> = {}
+  Object.entries(parameters).forEach(([k, v]) => {
+    const isArray = Array.isArray(v)
+    const hasTomlSchema = k in tomlSchema
+    const isIndexed = hasTomlSchema && tomlSchema[k].indexed === true
+    const isSectioned = hasTomlSchema && tomlSchema[k].sectioned === true
+    const isItemsSectioned = hasTomlSchema && tomlSchema[k].items?.sectioned === true
+    const hasNestedTomlSchema = hasTomlSchema && isObject(tomlSchema[k].items?.properties)
+    const nestedTomlSchema = hasNestedTomlSchema ? tomlSchema[k].items?.properties : {}
```

Comment:
I like these names, they really help to make the code readable.

Reply:
thanks